components/script/dom/datatransferitem.rs

```diff
@@ -0,0 +1,137 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use dom_struct::dom_struct;
+
+use crate::dom::bindings::callback::ExceptionHandling;
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::DataTransferItemBinding::{
+    DataTransferItemMethods, FunctionStringCallback,
+};
+use crate::dom::bindings::import::module::Rc;
+use crate::dom::bindings::reflector::{reflect_dom_object, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::bindings::weakref::MutableWeakRef;
+use crate::dom::datatransfer::DataTransfer;
+use crate::dom::file::File;
+use crate::dom::globalscope::GlobalScope;
+
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+pub enum Kind {
+    Text(DOMString),
+    File(DomRoot<File>),
```

Comment:
It's not _wrong_ to store a `DomRoot` inside of a dom_struct type, but `Dom` is more efficient. However, this requires one of two patterns for enums like this:
1. multiple enums (RootedKind and Kind) that store `DomRoot` and `Dom` respectively
2. using the enum for the actual storage and using Dom, and exposing multiple functions that accept different arguments (`new_text`, `new_file`)

I lean towards option 1 here, with one change:
```rust
pub enum Kind<'a> {
    Text(DOMString),
    File(&'a File),
}

enum KindStorage {
    Text(DOMString),
    File(Dom<File>),
}
```
Have the `new` and `new_inherited` methods take `Kind` arguments, and convert that into the appropriate `KindStorage` value when creating the actual DataTransferItem value.

Reply:
What should I do for `Dom<File>`error: Type must be rooted

Reply:
Add this attribute to the enum: https://github.com/servo/servo/blob/163e47766841725201f683495f4506f35fc9614c/components/script/dom/raredata.rs#L24